src/t_set.c

```diff
@@ -927,30 +1022,27 @@ void sinterGenericCommand(client *c, robj **setkeys,
      * the element against all the other sets, if at least one set does
      * not include the element it is discarded */
     si = setTypeInitIterator(sets[0]);
-    while((encoding = setTypeNext(si,&elesds,&intobj)) != -1) {
+    while((encoding = setTypeNext(si, &str, &len, &intobj)) != -1) {
         for (j = 1; j < setnum; j++) {
             if (sets[j] == sets[0]) continue;
-            if (encoding == OBJ_ENCODING_INTSET) {
-                /* intset with intset is simple... and fast */
-                if (sets[j]->encoding == OBJ_ENCODING_INTSET &&
-                    !intsetFind((intset*)sets[j]->ptr,intobj))
-                {
+            if (str == NULL && sets[j]->encoding == OBJ_ENCODING_INTSET) {
+                /* integer with intset is simple... and fast */
+                if (!intsetFind((intset*)sets[j]->ptr,intobj))
                     break;
-                /* in order to compare an integer with an object we
-                 * have to use the generic function, creating an object
-                 * for this */
-                } else if (sets[j]->encoding == OBJ_ENCODING_HT) {
-                    elesds = sdsfromlonglong(intobj);
-                    if (!setTypeIsMember(sets[j],elesds)) {
-                        sdsfree(elesds);
-                        break;
-                    }
-                    sdsfree(elesds);
-                }
             } else if (encoding == OBJ_ENCODING_HT) {
+                /* In this case, str is actually an sds string. */
+                if (!setTypeIsMember(sets[j], (sds)str)) {
+                    break;
+                }
+            } else {
+                /* We have to use the generic function, creating an object
+                 * for this */
+                sds elesds = str ? sdsnewlen(str, len) : sdsfromlonglong(intobj);
                 if (!setTypeIsMember(sets[j],elesds)) {
+                    sdsfree(elesds);
                     break;
                 }
+                sdsfree(elesds);
```

Comment:
maybe we can improve.
in case the other set is listpack, there shouldn't be a problem searching with char* and size_t.
that part seems easy to improve here.

in case the other set is a dict, in theory maybe we can create a dict type that supports non-sds lookup.
i'm not sure that part is worth it, but i do hate to see sds creation just for dictFind.
let's think about it.

Reply:
You mean we leave it for later? Then I agree. There are plenty of places where an sds string is created just for a dict lookup. It deserves a PR of its own, perhaps in combination with other changes to dict which are discussed in various places.